File: rank.go
package pinochle

import (
	"slices"
)

// -----------------------------------------------------------------------
// Type definitions
// -----------------------------------------------------------------------

// A rank is one of the six levels of cards in a pinochle deck.
type Rank int

const (
	NINE Rank = iota + 9
	JACK
	QUEEN
	KING
	TEN
	ACE
)

var offsetMap = map[Rank]int{
	NINE:  9,
	JACK:  11,
	QUEEN: 13,
	KING:  14,
	TEN:   10,
	ACE:   1,
}

var shortNameMap = map[Rank]string{
	NINE:  "9",
	JACK:  "J",
	QUEEN: "Q",
	KING:  "K",
	TEN:   "10",
	ACE:   "A",
}

var rankNameMap = map[Rank]string{
	NINE:  "9",
	JACK:  "jack",
	QUEEN: "queen",
	KING:  "king",
	TEN:   "10",
	ACE:   "ace",
}

// Ranks is a slice of the six ranks in ascending order of precedence
var Ranks = []Rank{NINE, JACK, QUEEN, KING, TEN, ACE}

// -----------------------------------------------------------------------
// Methods
// -----------------------------------------------------------------------

// Less returns true if this rank is less than that of the other
func (r Rank) Less(other Rank) bool {
	return slices.Index(Ranks, r) < slices.Index(Ranks, other)
}

// Returns the offset in the Unicode system to the beginning of cards
// for this rank
func (r Rank) Offset() int {
	return offsetMap[r]
}

// Returns a representation of a Rank as a string
func (r Rank) String() string {
	return shortNameMap[r]
}

File: src/cards/rank.py
from enum import Enum


class Rank(Enum):
    NINE = (1, "9", "9")
    JACK = (2, "J", "jack")
    QUEEN = (3, "Q", "queen")
    KING = (4, "K", "king")
    TEN = (5, "10", "10")
    ACE = (6, "A", "ace")

    def __init__(self, value, short_name, rank_name):
        self._value_ = value
        self._short_name = short_name
        self._rank_name = rank_name

    @property
    def short_name(self):
        return self._short_name

    @property
    def rank_name(self):
        return self._rank_name
